src/controllers/products/recommendedController.js

```diff
@@ -0,0 +1,263 @@
+    }
+    res.status(200).json({
+      success: true,
+      message: `${req.t('Updated product succesfully ')} ${req.params.id}`,
+      response: products,
+    });
+  } catch (error) {
+    res.status(500).json({
+      success: false,
+      message: `${req.t('Update product Error')} ${req.params.id}`,
+      Error: error.message,
+    });
+  }
+};
+// function to delete product by their ID.
+export const deleteproduct = async (req, res) => {
+  try {
+    const { authenticatedUser } = req;
+    if (!authenticatedUser) {
+      return res.status(403).json({
+        success: false,
+        message: req.t('unauthorized_msg'),
+      });
+    }
+    const products = await product.findByPk(req.params.id);
+    await product.destroy({
+      where: {
+        id: req.params.id,
+      },
+    });
+    if (products === null) {
+      res.status(404).json({ success: false, message: req.t('Not Found!! ') });
+    }
+    res.status(200).json({
+      success: true,
+      message: `${req.t('Deleted Successfully! ')} ${req.params.id}`,
+      response: products,
+    });
+  } catch (error) {
+    res.status(500).json({
+      success: false,
+      message: `${req.t('Delete product Error ')} ${req.params.id}`,
+      Error: error.message,
+    });
+  }
+};
+
+async function getRecommendedProducts(userID) {
+  try {
+    // Get the list of products that the buyer has viewed
+    const viewedProducts = await productViews.findAll({
+      where: {
+        buyerId: userID,
+      },
+      attributes: ['productId'],
+      raw: true,
+    });
+    const viewedProductIds = viewedProducts.map((viewedProduct) => viewedProduct.productId);
+    // Find other buyers who have viewed the same products
+    const similarBuyers = await productViews.findAll({
+      where: {
+        productId: viewedProductIds,
+        buyerId: { [Op.not]: userID },
+      },
+      attributes: ['buyerId'],
+      group: ['buyerId'],
+      raw: true,
+    });
+    const similarBuyerIds = similarBuyers.map((similarBuyer) => similarBuyer.buyerId);
+    // Find the products that these similar buyers have viewed
+    const recommendedProducts = await productViews.findAll({
+      where: {
+        buyerId: similarBuyerIds,
+        productId: { [Op.notIn]: viewedProductIds },
+      },
+      attributes: ['productId'],
+      group: ['productId'],
+      raw: true,
+    });
+    const recommendedProductIds = recommendedProducts.map(
+      (recommendedProduct) => recommendedProduct.productId,
+    );
+    // Get the details of the recommended products
+    const finalRecommendedProducts = await product.findAll({
+      where: { id: recommendedProductIds },
+      raw: true,
+    });
+    return finalRecommendedProducts;
+  } catch (error) {
+    console.log(error);
+    return error;
+  }
+}
+// function to get recommended product
+export const RecommendedProduct = async (req, res) => {
```

Comment:
I have tested this function and it does well when a user has viewed products similar to other users as it goes ahead to recommend other products viewed by those users. However, when the user has not viewed an product similar to other users they don't get any recommendation. I would suggest that in that case, you recommend products in the same category as the product the user has already viewed. For instance, a user viewed a jacket but no other users viewed that jacket, so they could get recommendations of other products in the fashion category as it seems to be a category of interest.

Reply:
Yes i'm still working to add other parameters while generating the recommended products. I will also include that.